# Assignment 1: Happiness
### Name: Vinay Haresh Bojja
#### Instruction:
* Install a library using the command: pip3 install xlrd.

In [1]:
import pandas as pd

# Read Excel file 
df = pd.read_excel('DataForTable2.1WHR2023.xls')
df

/var/folders/fq/nclhl23j55s570qj6dfkqgq80000gn/T/ipykernel_56266/3673298581.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Country name  year  Life Ladder  Log GDP per capita  Social support  \
0     Afghanistan  2008     3.723590            7.350416        0.450662   
1     Afghanistan  2009     4.401778            7.508646        0.552308   
2     Afghanistan  2010     4.758381            7.613900        0.539075   
3     Afghanistan  2011     3.831719            7.581259        0.521104   
4     Afghanistan  2012     3.782938            7.660506        0.520637   
...           ...   ...          ...                 ...             ...   
2194     Zimbabwe  2018     3.616480            7.783066        0.775388   
2195     Zimbabwe  2019     2.693523            7.697755        0.759162   
2196     Zimbabwe  2020     3.159802            7.596050        0.717243   
2197     Zimbabwe  2021     3.154578            7.656878        0.685151   
2198     Zimbabwe  2022     3.296220            7.670123        0.666172   

      Healthy life expectancy at birth  Freedom to make life choices  \
0                            50.500000                      0.718114   
1                            50.799999                      0.678896   
2                            51.099998                      0.600127   
3                            51.400002                      0.495901   
4                            51.700001                      0.530935   
...                                ...                           ...   
2194                         52.625000                      0.762675   
2195                         53.099998                      0.631908   
2196                         53.575001                      0.643303   
2197                         54.049999                      0.667636   
2198                         54.525002                      0.651987   

      Generosity  Perceptions of corruption  Positive affect  Negative affect  
0       0.167652                   0.881686         0.414297         0.258195  
1       0.190809                   0.850035         0.481421         0.237092  
2       0.121316                   0.706766         0.516907         0.275324  
3       0.163571                   0.731109         0.479835         0.267175  
4       0.237588                   0.775620         0.613513         0.267919  
...          ...                        ...              ...              ...  
2194   -0.051219                   0.844209         0.657524         0.211726  
2195   -0.047464                   0.830652         0.658434         0.235354  
2196    0.006313                   0.788523         0.660658         0.345736  
2197   -0.075575                   0.756945         0.609917         0.241682  
2198   -0.069513                   0.752632         0.640609         0.191350  

[2199 rows x 11 columns]

1. What is your evaluation of the quality of the data? How large is the dataset? What are the
features (columns) and how are they represented? Are there any interesting distributions in the
(numeric) features? Are there any interesting correlations in the data? Can you identify any
limitations, missing values or distortions of the data? What would you like to see in this dataset?

In [6]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2199 entries, 0 to 2198
Data columns (total 11 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Country name                      2199 non-null   object 
 1   year                              2199 non-null   int64  
 2   Life Ladder                       2199 non-null   float64
 3   Log GDP per capita                2179 non-null   float64
 4   Social support                    2186 non-null   float64
 5   Healthy life expectancy at birth  2145 non-null   float64
 6   Freedom to make life choices      2166 non-null   float64
 7   Generosity                        2126 non-null   float64
 8   Perceptions of corruption         2083 non-null   float64
 9   Positive affect                   2175 non-null   float64
 10  Negative affect                   2183 non-null   float64
dtypes: float64(9), int64(1), object(1)
memory usage: 189.1+ KB


#### Question 1.a) What is your evaluation of the quality of the data?
#### Answer 1.a)
 The data has in total 2199 rows. Some columns such as "Log GDP per capita", "Social support", "Healthy life expectancy at birth", "Freedom to make life choice", "Generosity", "Perceptions of corruption", "Positive affect" and "Negative affect" have null values in them. Hence, I think the data needs to be cleaned. 

In [7]:
df

Country name  year  Life Ladder  Log GDP per capita  Social support  \
0     Afghanistan  2008     3.723590            7.350416        0.450662   
1     Afghanistan  2009     4.401778            7.508646        0.552308   
2     Afghanistan  2010     4.758381            7.613900        0.539075   
3     Afghanistan  2011     3.831719            7.581259        0.521104   
4     Afghanistan  2012     3.782938            7.660506        0.520637   
...           ...   ...          ...                 ...             ...   
2194     Zimbabwe  2018     3.616480            7.783066        0.775388   
2195     Zimbabwe  2019     2.693523            7.697755        0.759162   
2196     Zimbabwe  2020     3.159802            7.596050        0.717243   
2197     Zimbabwe  2021     3.154578            7.656878        0.685151   
2198     Zimbabwe  2022     3.296220            7.670123        0.666172   

      Healthy life expectancy at birth  Freedom to make life choices  \
0                            50.500000                      0.718114   
1                            50.799999                      0.678896   
2                            51.099998                      0.600127   
3                            51.400002                      0.495901   
4                            51.700001                      0.530935   
...                                ...                           ...   
2194                         52.625000                      0.762675   
2195                         53.099998                      0.631908   
2196                         53.575001                      0.643303   
2197                         54.049999                      0.667636   
2198                         54.525002                      0.651987   

      Generosity  Perceptions of corruption  Positive affect  Negative affect  
0       0.167652                   0.881686         0.414297         0.258195  
1       0.190809                   0.850035         0.481421         0.237092  
2       0.121316                   0.706766         0.516907         0.275324  
3       0.163571                   0.731109         0.479835         0.267175  
4       0.237588                   0.775620         0.613513         0.267919  
...          ...                        ...              ...              ...  
2194   -0.051219                   0.844209         0.657524         0.211726  
2195   -0.047464                   0.830652         0.658434         0.235354  
2196    0.006313                   0.788523         0.660658         0.345736  
2197   -0.075575                   0.756945         0.609917         0.241682  
2198   -0.069513                   0.752632         0.640609         0.191350  

[2199 rows x 11 columns]

#### Question 1.b) How large is the dataset? 
#### Answer 1.b) 
* The dataset has 2199 rows and 11 columns.

#### Question 1.c) What are the features (columns) and how are they represented?
#### Answer 1.c) 
* The features are as listed below:
    * Country name -
       * DataType - object
       * Desc - Name of the country.
    * year -
       * Datatype - int64 
       * Desc - Year the data was collected from.
    * Life Ladder -
       * Datatype - float64 
       * Desc - Imagine a ladder with steps in a range of 0 to 10, with 0 being the worst and 10 being the best possible life.
    * Log GDP per capita -
       * Datatype - float64
       * Desc - GDP of the country for that particular year.
    * Social support -
       * Datatype - float64
       * Desc - Average of responses(0s or 1s) from the people, when asked whether they had a relative or a friend to help when they needed them in trouble. 
    * Healthy life expectancy at birth -
       * Datatype - float64
       * Desc - Mean of number of years a person is expected to live in full health at birth.
    * Freedom to make life choices -
       * Datatype - float64
       * Desc - Average of responses(0s or 1s) from the people, when asked whether they were satisfied with the freedom to choose what they want in their life.
    * Generosity -
       * Datatype - float64
       * Desc - Average of responses(0s or 1s) from the people, when asked whether they donated money to the charity in the last month.
    * Perceptions of corruption -
       * Datatype - float64
       * Desc - Average of responses(0s or 1s) from the people, when asked whether corruption is widespread throughout the government or not, or if it was widespread throughout the businesses or not.
    * Positive affect -
       * Datatype - float64
       * Desc - Average of responses(0s or 1s) from the people, when asked whether they laughed, enjoyed, or did some interesting           things.
    * Negative affect -
       * Datatype - float64
       * Desc - Average of responses (0s or 1s) from the people upon asking whether they were worried, sad, or angry.
(Reference: John F. Helliwell, Haifang Huang, Max Norton, Shun Wang and Leonard Gof. https://happiness-report.s3.amazonaws.com/2023/WHR+23_Statistical_Appendix.pdf. 1-2)   

In [8]:
df.describe()

year  Life Ladder  Log GDP per capita  Social support  \
count  2199.000000  2199.000000         2179.000000     2186.000000   
mean   2014.161437     5.479226            9.389766        0.810679   
std       4.718736     1.125529            1.153387        0.120952   
min    2005.000000     1.281271            5.526723        0.228217   
25%    2010.000000     4.646750            8.499764        0.746609   
50%    2014.000000     5.432437            9.498955        0.835535   
75%    2018.000000     6.309460           10.373216        0.904792   
max    2022.000000     8.018934           11.663788        0.987343   

       Healthy life expectancy at birth  Freedom to make life choices  \
count                       2145.000000                   2166.000000   
mean                          63.294583                      0.747858   
std                            6.901104                      0.140150   
min                            6.720000                      0.257534   
25%                           59.119999                      0.656528   
50%                           65.050003                      0.769821   
75%                           68.500000                      0.859382   
max                           74.474998                      0.985178   

        Generosity  Perceptions of corruption  Positive affect  \
count  2126.000000                2083.000000      2175.000000   
mean      0.000096                   0.745195         0.652143   
std       0.161083                   0.185837         0.105922   
min      -0.337527                   0.035198         0.178886   
25%      -0.112116                   0.688139         0.571684   
50%      -0.022671                   0.799654         0.663063   
75%       0.092070                   0.868827         0.737936   
max       0.702708                   0.983276         0.883586   

       Negative affect  
count      2183.000000  
mean          0.271501  
std           0.086875  
min           0.082737  
25%           0.207660  
50%           0.260671  
75%           0.322894  
max           0.704590

#### Question 1.d) Are there any interesting distributions in the (numeric) features? 
#### Answer 1.d)
*
    * Life Ladder - The countries exhibit a range of life ladder, with the minimum being 1.28, the maximum reaching 8.01, and the average settling at 5.47. This suggests that countries experience a balance between relatively favorable and challenging life circumstances.
    * Log GDP per capita - Highest GDP observed is 11.66 while the lowest is 5.52.
    * Social support - Around 81% countries had social support from their relatives or friends when in trouble.
    * Freedom to make life choices - On an average, 74% of countries had the freedom to make life choices with the max being 98%.
    * Generosity - *********
    * Perception of corruption - In the country, approximately 3% of the population believes that corruption is widespread, while the highest percentage of people who perceive the country as corruption-free is around 98%.
    * Positive affect - On average, 65% of the countries engaged in laughter, enjoyment, and activities they found interesting.
    * Negative affect - In a country, only 8% of the population reported feeling worried, sad, or angry, indicating that it had the least negative impact on their lives.

#### Question 1.e) Are there any interesting correlations in the data?
#### Answer 1.e)

#### Question 1.f) Can you identify any limitations, missing values or distortions of the data? 
#### Answer 1.f)

#### Question 1.g What would you like to see in this dataset?
#### Answer 1.g)

2. What is the happiest country in 2023? In 2013? Of all time? What is the least happy country in
2023, 2013 and of all time?

3. Of the countries in this dataset, which country had the largest increase in happiness from its start
of participation to 2023? Which had the largest decrease in happiness?